contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,781 @@
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namepace when
+-  using schedulers,
+-  placing pods on specific nodes (master roles for instance),
+-  using specific priority classes,
+-  expressing pod affinity or anti-affinity rules.
+
+Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+Implementing SchedulingPolicy implies:
+-  Creating a new resource named **SchedulingPolicy** (schedpol)
+-  Creating an **AdmissionController** that dehaves on a deny-all-but basis
+-  Allow SchedulingPolicy to be used by pods using RoleBindings or ClusterRoleBindings
+
+# Detailed Design
+
+SchedulingPolicy resource specs are composed of several main attributes:
+-  **Required** scheduling components. These list the manadory NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations and optionally valid values that have to be provided to allow scheduling.
+-  **Allowed** scheduling components. These list the optional components that can be specified in pods definition.
+-  **Default** scheduling components. These list default values to set unless specified in pods definition.
+
+SchedulingPolicy resources are supposed to apply in a deny-all-except approach.
+
+An AdmissionController must be added to the mutating phase to
+-  add default values if unspecified for NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations,
+-  reject pod scheduling if the serviceaccount running the pod is not allowed to specify requested NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations.
+
+All usable scheduling policies (allowed by RBAC) are merged before evaluating if scheduling constraints defined in pods are allowed.
+
+## SchedulingPolicy
+
+Proposed API group: `extensions/v1alpha1`
+
+SchedulingPolicy is a cluster-scoped resource (not namespaced).
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this schedpol won't allow any item from the missing field.
+
+```yaml
+apiVersion: extensions/valpha1
+kind: SchedulingPolicy
+metadata:
+  name: my-schedpol
+spec:
+  required:
+    schedulerNames:      # Describes schedulers names that are required
+    priorityClasseNames: # Describes priority class names that are required
+    nodeSelectors:       # Describes node selectors that must be used
+    affinities:          # Describes affinities that must be used
+  allowed:
+    schedulerNames:      # Describes schedulers names that are allowed
```

Comment:
An option to consider is to create a sub-type called "SchedulingPolicyItems" (or something like that) and place
"schedulerNames", "priorityClassNames", ... under that sub-type. Each one of the "required", "allowed", "default", and "forbidden" will of type "SchedulingPolicyItems". Then in the validation logic you can check which one of the fields in "schedulingPolicyItems" can be/must be present or absent or how many elements they can have. For example, "default.schedulerNames" can have only one item.

Reply:
Do we expect different policies to apply to different workloads? I'd assume so. We should think about how that should be specified.

For example:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/metadata-policy.md
https://kubernetes.io/docs/concepts/services-networking/network-policies/

Do we expect multiple polices to apply to the same pod? When/where/why? Should they be non-overlapping? Example scenarios would help.